File: baba-chatbot-python/src/baba_chatbot/application/conversation/orchestrator/conversation_orchestrator.py
"""
Orquestador de conversaciones que gestiona el flujo completo de validaciones,
generación de respuestas y sistema de reincidencia.
"""
import logging
from typing import List, Optional, Dict
from dataclasses import dataclass, field
from ...abstractions.interfaces import ILlmClient, IGuardrailsValidator
from ..guardrails.content_moderation_result import ModerationFlag, ModerationSeverity


logger = logging.getLogger(__name__)


@dataclass
class ConversationResponse:
    """
    Respuesta del orquestador de conversación.
    """
    success: bool
    message: str
    requires_escalation: bool = False
    moderation_flags: List[ModerationFlag] = field(default_factory=list)
    has_pii_violations: bool = False

    def __post_init__(self):
        """Asegura que moderation_flags sea una lista mutable."""
        if not isinstance(self.moderation_flags, list):
            self.moderation_flags = list(self.moderation_flags) if self.moderation_flags else []


class ConversationOrchestrator:
    """
    Orquesta el flujo completo de conversación con validaciones, generación de respuestas
    y gestión de violaciones (sistema de reincidencia de 3 strikes).
    """

    def __init__(
        self,
        llm_client: ILlmClient,
        guardrails_validator: IGuardrailsValidator
    ):
        """
        Inicializa el orquestador de conversaciones.
        
        Args:
            llm_client: Cliente para comunicación con el LLM
            guardrails_validator: Validador de guardrails
        """
        self._llm_client = llm_client
        self._guardrails_validator = guardrails_validator
        self._user_violation_count: Dict[str, int] = {}

    async def process_message(
        self,
        user_id: str,
        user_message: str,
        system_prompt: str,
        conversation_history: Optional[List[str]] = None
    ) -> ConversationResponse:
        """
        Procesa un mensaje del usuario a través del pipeline completo de validación,
        generación de respuesta y verificación de guardrails.
        
        Args:
            user_id: Identificador del usuario
            user_message: Mensaje del usuario
            system_prompt: Prompt del sistema con instrucciones
            conversation_history: Historial de conversación opcional
            
        Returns:
            Respuesta procesada con el resultado de todas las validaciones
        """
        logger.info(f"Processing message for user {user_id}")

        # Validar entrada del usuario
        user_validation = self._guardrails_validator.validate_user_input(user_message)

        if not user_validation.is_appropriate:
            logger.warning(
                f"User message failed validation for user {user_id}. "
                f"Flags: {[f.value for f in user_validation.flags]}"
            )

            # Incrementar contador de violaciones
            violation_count = self._user_violation_count.get(user_id, 0) + 1
            self._user_violation_count[user_id] = violation_count

            # Determinar si se debe terminar la conversación
            if (user_validation.severity == ModerationSeverity.HIGH or
                    violation_count >= 3):
                logger.warning(
                    f"User {user_id} has {violation_count} violations. "
                    "Terminating conversation."
                )

                return ConversationResponse(
                    success=False,
                    message=self._guardrails_validator.get_response_for_violation(
                        user_validation,
                        violation_count
                    ),
                    requires_escalation=True,
                    moderation_flags=user_validation.flags
                )

            return ConversationResponse(
                success=False,
                message=user_validation.suggested_response or (
                    "Lo siento, no puedo ayudarte con eso. "
                    "¿Hay algo relacionado con vehículos en lo que pueda asistirte?"
                ),
                requires_escalation=False,
                moderation_flags=user_validation.flags
            )

        # Generar respuesta del LLM
        try:
            llm_response = await self._llm_client.generate_response(
                system_prompt,
                user_message,
                conversation_history
            )
        except Exception as ex:
            logger.error(f"Error generating LLM response for user {user_id}: {ex}")
            
            return ConversationResponse(
                success=False,
                message="Lo siento, tuve un problema técnico. ¿Podrías intentar de nuevo?",
                requires_escalation=False
            )

        # Validar respuesta del LLM
        response_validation = await self._guardrails_validator.validate_response(llm_response)

        if not response_validation.is_valid:
            logger.error(
                f"LLM response failed guardrails validation for user {user_id}. "
                f"Violations: {response_validation.violations}"
            )

            if response_validation.requires_human_escalation:
                return ConversationResponse(
                    success=False,
                    message=(
                        "Déjame conectarte con un asesor especializado que podrá ayudarte mejor. "
                        "Dame un momento..."
                    ),
                    requires_escalation=True
                )

            return ConversationResponse(
                success=False,
                message=(
                    "Disculpa, déjame reformular eso. "
                    "¿Podrías ser más específico sobre lo que necesitas?"
                ),
                requires_escalation=False
            )

        # Usar contenido limpio (con PII enmascarada)
        final_response = response_validation.cleaned_content or llm_response

        # Validar calidad de la respuesta
        if not self._guardrails_validator.validate_response_quality(final_response):
            logger.warning(f"Response failed quality validation for user {user_id}")
            
            return ConversationResponse(
                success=False,
                message=(
                    "¿Podrías darme más detalles sobre lo que buscas? "
                    "Así puedo ayudarte mejor."
                ),
                requires_escalation=False
            )

        logger.info(f"Successfully processed message for user {user_id}")

        # Limpiar contador de violaciones en caso de éxito
        if user_id in self._user_violation_count:
            del self._user_violation_count[user_id]

        # Verificar si hubo violaciones de PII
        has_pii_violations = any(
            "detected" in v.lower()
            for v in response_validation.violations
        )

        return ConversationResponse(
            success=True,
            message=final_response,
            requires_escalation=False,
            has_pii_violations=has_pii_violations
        )

    def reset_violation_count(self, user_id: str) -> None:
        """
        Reinicia el contador de violaciones para un usuario específico.
        
        Args:
            user_id: Identificador del usuario
        """
        if user_id in self._user_violation_count:
            del self._user_violation_count[user_id]

File: baba-chatbot-net/src/Baba.Chatbot.Application/Conversation/Orchestrator/ConversationOrchestrator.cs
using Baba.Chatbot.Application.Abstractions;
using Baba.Chatbot.Application.Conversation.Guardrails;
using Microsoft.Extensions.Logging;
using System.Collections.Concurrent;

namespace Baba.Chatbot.Application.Conversation.Orchestrator;

public class ConversationOrchestrator
{
    private readonly ILlmClient _llmClient;
    private readonly IGuardrailsValidator _guardrailsValidator;
    private readonly ILogger<ConversationOrchestrator> _logger;
    
    private readonly ConcurrentDictionary<string, int> _userViolationCount = new();

    public ConversationOrchestrator(ILlmClient llmClient, IGuardrailsValidator guardrailsValidator, ILogger<ConversationOrchestrator> logger)
    {
        _llmClient = llmClient;
        _guardrailsValidator = guardrailsValidator;
        _logger = logger;
    }

    public async Task<ConversationResponse> ProcessMessageAsync(string userId, string userMessage, string systemPrompt, List<string>? conversationHistory = null, CancellationToken cancellationToken = default)
    {
        _logger.LogInformation("Processing message for user {UserId}", userId);

        var userValidation = _guardrailsValidator.ValidateUserInput(userMessage);

        if (!userValidation.IsAppropriate)
        {
            _logger.LogWarning("User message failed validation for user {UserId}. Flags: {Flags}",
                userId, string.Join(", ", userValidation.Flags));


            var violationCount = _userViolationCount.AddOrUpdate(userId, 1, (key, count) => count + 1);

     
            if (userValidation.Severity == ModerationSeverity.High || violationCount >= 3)
            {
                _logger.LogWarning("User {UserId} has {ViolationCount} violations. Terminating conversation.", 
                    userId, violationCount);

                return new ConversationResponse
                {
                    Success = false,
                    Message = _guardrailsValidator.GetResponseForViolation(userValidation, violationCount),
                    RequiresEscalation = true,
                    ModerationFlags = userValidation.Flags
                };
            }

            return new ConversationResponse
            {
                Success = false,
                Message = userValidation.SuggestedResponse ?? 
                         "Lo siento, no puedo ayudarte con eso. ¿Hay algo relacionado con vehículos en lo que pueda asistirte?",
                RequiresEscalation = false,
                ModerationFlags = userValidation.Flags
            };
        }

        string llmResponse;
        try
        {
            llmResponse = await _llmClient.GenerateResponseAsync(systemPrompt, userMessage, conversationHistory, cancellationToken);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error generating LLM response for user {UserId}", userId);
            
            return new ConversationResponse
            {
                Success = false,
                Message = "Lo siento, tuve un problema técnico. ¿Podrías intentar de nuevo?",
                RequiresEscalation = false
            };
        }

        var responseValidation = await _guardrailsValidator.ValidateResponseAsync(llmResponse);

        if (!responseValidation.IsValid)
        {
            _logger.LogError("LLM response failed guardrails validation for user {UserId}. Violations: {Violations}", userId, string.Join(", ", responseValidation.Violations));

            if (responseValidation.RequiresHumanEscalation)
            {
                return new ConversationResponse
                {
                    Success = false,
                    Message = "Déjame conectarte con un asesor especializado que podrá ayudarte mejor. Dame un momento...",
                    RequiresEscalation = true
                };
            }

            return new ConversationResponse
            {
                Success = false,
                Message = "Disculpa, déjame reformular eso. ¿Podrías ser más específico sobre lo que necesitas?",
                RequiresEscalation = false
            };
        }

        var finalResponse = responseValidation.CleanedContent ?? llmResponse;

        if (!_guardrailsValidator.ValidateResponseQuality(finalResponse))
        {
            _logger.LogWarning("Response failed quality validation for user {UserId}", userId);
            
            return new ConversationResponse
            {
                Success = false,
                Message = "¿Podrías darme más detalles sobre lo que buscas? Así puedo ayudarte mejor.",
                RequiresEscalation = false
            };
        }

        _logger.LogInformation("Successfully processed message for user {UserId}", userId);

        if (_userViolationCount.ContainsKey(userId))
        {
            _userViolationCount.TryRemove(userId, out _);
        }

        return new ConversationResponse
        {
            Success = true,
            Message = finalResponse,
            RequiresEscalation = false,
            HasPiiViolations = responseValidation.Violations.Any(v => v.Contains("detected"))
        };
    }


    public void ResetViolationCount(string userId)
    {
        _userViolationCount.TryRemove(userId, out _);
    }
}

public class ConversationResponse
{
    public bool Success { get; set; }
    public required string Message { get; set; }
    public bool RequiresEscalation { get; set; }
    public List<ModerationFlag> ModerationFlags { get; set; } = new();
    public bool HasPiiViolations { get; set; }
}
